dev/ToastNotifications/ToastNotificationUtility.cpp

```diff
@@ -0,0 +1,190 @@
+    DWORD bufferLength = sizeof(registeredGuidBuffer);
+    HRESULT status = RegGetValueW(
+        hKey.get(),
+        nullptr,
+        L"ToastGUID",
+        RRF_RT_REG_SZ,
+        nullptr,
+        &registeredGuidBuffer,
+        &bufferLength);
+
+    if (status == ERROR_FILE_NOT_FOUND)
+    {
+        GUID newToastGuid;
+        THROW_IF_FAILED(CoCreateGuid(&newToastGuid));
+
+        wil::unique_cotaskmem_string newToastGuidString;
+        THROW_IF_FAILED(StringFromCLSID(newToastGuid, &newToastGuidString));
+
+        std::wstring guidWideStr{ newToastGuidString.get() };
+        RegisterValue(hKey, L"ToastGUID", reinterpret_cast<const BYTE*>(guidWideStr.c_str()), REG_SZ, guidWideStr.size() * sizeof(wchar_t));
+        return guidWideStr;
+    }
+    else
+    {
+        THROW_HR_IF(status, FAILED_WIN32(status));
+    }
+    return registeredGuidBuffer;
+}
+
+std::wstring RetrieveToastGuid()
+{
+    if (AppModel::Identity::IsPackagedProcess())
+    {
+        wchar_t appUserModelId[APPLICATION_USER_MODEL_ID_MAX_LENGTH] = {};
+        UINT32 appUserModelIdSize{ APPLICATION_USER_MODEL_ID_MAX_LENGTH };
+
+        THROW_IF_FAILED(GetCurrentApplicationUserModelId(&appUserModelIdSize, appUserModelId));
+        return appUserModelId;
+    }
+    else
+    {
+        return RetrieveUnpackagedToastGuid();
+    }
+}
+
+void RegisterAssets(std::wstring const& appId, winrt::Microsoft::Windows::ToastNotifications::ToastAssets const& activationInfo, wil::unique_cotaskmem_string const& clsid)
+{
+    wil::unique_hkey hKey;
+    std::wstring subKey{ c_appIdentifierPath + appId };
+
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr,
+        &hKey,
+        nullptr));
+
+    RegisterValue(hKey, L"DisplayName", reinterpret_cast<const BYTE*>(activationInfo.DisplayName().c_str()), REG_EXPAND_SZ, activationInfo.DisplayName().size() * sizeof(wchar_t));
+    RegisterValue(hKey, L"IconUri", reinterpret_cast<const BYTE*>(activationInfo.IconPath().AbsoluteUri().c_str()), REG_EXPAND_SZ, activationInfo.IconPath().AbsoluteUri().size() * sizeof(wchar_t));
+
+    std::wstring wideStringClsid{ clsid.get() };
+    RegisterValue(hKey, L"CustomActivator", reinterpret_cast<const BYTE*>(wideStringClsid.c_str()), REG_SZ, wideStringClsid.size() * sizeof(wchar_t));
+}
+
+void RegisterComServer(wil::unique_cotaskmem_string const& processName, wil::unique_cotaskmem_string const& clsid)
+{
+    wil::unique_hkey hKey;
+    std::wstring subKey{ c_clsIdPath + clsid.get() + L"\\LocalServer32" };
+
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr,
+        &hKey,
+        nullptr));
+
+    std::wstring comRegistrationExeString{ L"\"" };
```

Comment:
I was taught this cool new trick to avoid \"

std::wstring comRegistrationExeString{ LR(""") };